### ISSUE TITLE
Postgres table names with special characters truncated when specifying the TABLES open option

### ISSUE THREAD

**Issue by arredond** [OSGeo/gdal#11486]
### What is the bug?

Postgres table names may contain special characters such as parentheses or square and curly brackets. However, when specifying one or more of these tables using the `TABLES` open option, the layer name used by GDAL is truncated (up to the first invalid special character) and all subsequent queries performed by GDAL fail.

I have not found a way to escape these special characters, be it via single/double quotes, backticks, or using functionality such as Python's `shlex.quote`.

Note: I initially ran into this problem when specifying open options using the Python bindings (`osgeo.gdal.OpenEx(url, open_options=["TABLES=..."]`) but the error also occurs when calling GDAL directly via console utilities, such as `ogrinfo`

### Steps to reproduce the issue

Copy the following snippet in a terminal to list a specific table from a Supabase database using a read-only user:

```bash
ogrinfo -oo ACTIVE_SCHEMA=public -oo "TABLES=null_island_points (WITH PARENS)" -so -al "postgresql://read_only_user.amcliteeqmfizqcmlmpf:read_only_pwd@aws-0-us-east-1.pooler.supabase.com:6543/postgres"
```

Several errors appear, all of them due to the fact that GDAL has interpreted the table name as `null_island_points ` (with trailing space)

```
Layer name: public.null_island_points
Geometry: Unknown (any)
ERROR 1: ERROR:  relation "public.null_island_points " does not exist
LINE 1: SELECT count(*) FROM "public"."null_island_points "
                             ^

Feature Count: 0
ERROR 1: ERROR:  relation "public.null_island_points " does not exist
LINE 1: ...r0x128e06ca0 CURSOR for SELECT "WITH PARENS" FROM "public"."...
```

### Versions and provenance

- GDAL version: GDAL 3.10.0, released 2024/11/01
- Operating system: macOS Sonoma 14.5 (23F79)
- Installed via [homebrew](https://formulae.brew.sh/formula/gdal)
- Postgres version: PostgreSQL 15.6 on aarch64-unknown-linux-gnu, compiled by gcc (GCC) 13.2.0, 64-bit
- PostGIS version: 3.3 USE_GEOS=1 USE_PROJ=1 USE_STATS=1

**Comment by jratike80** [comment]
Have you tried
`-oo "TABLES=\"null_island_points (WITH PARENS)\""`

Alternative you can try `-sql "select * from \"null_island_points (WITH PARENS)\""`

In SQL an that kind of identifier must be enclosed between double quotation marks, and for the GDAL command line the " must be escaped if they appear in the middle of outes "s.

**Comment by rouault** [comment]
This indeed cannot work in the current state of the driver. The syntax "TABLES=foo(bar)" is a bit of legacy at a time where the driver only supported one geometry column per layer, and was meant as "table foo, using geometry column bar".

I'm not totally sure what syntax we could use  to really mean table foo(bar).  Perhaps backslash escape the paranthesis ?  "TABLES=foo\(bar\)" ? which will probably require escaping the backslash themselves in most shells and programming languages...

**Comment by arredond** [comment]
Oh interesting, I wasn't aware that the parentheses had special meaning (specifying a geometry column). Are there other special characters? Square or curly brackets, perhaps?

As for what syntax to use: I find backslash-escaping the parenthesis intuitive (and indeed was the first thing I tried). Even if that means escaping the backslash as well, I think that's okay, as long as it's well documented.

An alternative, as @jratike80 points out, would be to enclose the table names in double quotes (which I also tried)

**Comment by jratike80** [comment]
"would be to enclose the table names in double quotes (which I also tried)"

I suppose that you mean that the trial was not successfull.

**Comment by rouault** [comment]
> Are there other special characters? Square or curly brackets, perhaps?

no, all answers are at https://github.com/OSGeo/gdal/blob/master/ogr/ogrsf_frmts/pg/ogrpgdatasource.cpp#L923 . So it first tokenizes on "," as the separator between tables, then "." to separate schema from table, and then "(" for table name and geometry column

> An alternative, as @jratike80 points out, would be to enclose the table names in double quotes (which I also tried)

No alternatives. Requires a code change.

**Comment by jratike80** [comment]
`-sql "select...` seems to work. 

```
ogr2ogr pg:"host=localhost port=5432 ..." states.shp -nln "null_island_points (with parens)" -nlt multipolygon

ogrinfo pg:"host=localhost port=5432 ..." -sql "select * from \"null_island_points (with parens)\"" -so
INFO: Open of `pg:host=localhost port=5432 ...'
      using driver `PostgreSQL' successful.

Layer name: sql_statement
Geometry: Unknown (any)
Feature Count: 49
Extent: (-124.731422, 24.955967) - (-66.969849, 49.371735)
Layer SRS WKT:
```

**Comment by rouault** [comment]
fix in https://github.com/OSGeo/gdal/pull/11490
